src/SafeBase.jl

```diff
@@ -0,0 +1,165 @@
+`Base.parse`, as Result types in Julia are generally more performant than throwing exceptions
+and provide cleaner control flow and error reporting than the other solutions.
+"""
+function safe_parse end
+
+function safe_parse(T::Type{<:Union{<:AbstractFloat,<:Integer}}, slurp...; kwargs...)::Result{T,ParseError}
+  res = tryparse(T, slurp...; kwargs...)
+  isnothing(res) && return ParseError(T, slurp, "Could not parse $(slurp...) into type $T", backtrace())
+  res
+end
+
+
+# ---------------------------------------
+# # Julia parsing
+
+"""
+Alternative to `Meta.parse` based on `JuliaSyntax.jl` library. Function doesn't throw, but wraps returned value (or error) in a `Result`.
+As opposed to `Meta.parse`, `:error` and `:incomplete` expressions both are treated as `ParseError`.
+
+Optional argument `rule` can accept values :statement or :all, indicating parsing single-expression vs multi-expression inputs, respectively.
+Optional argument `filename` is passed to set any file name information, if applicable. This will also annotate errors and warnings with the
+source file name.
+
+Implementation of this function mimics [JuliaSyntax._parse](https://github.com/JuliaLang/JuliaSyntax.jl/blob/3bf262bba32e833ed6d0d59a455a62faee97b408/src/parser_api.jl#L77)
+"""
+function safe_parse_julia(
+  str::AbstractString,
+  rule::Symbol=:statement,
+  filename::Union{Nothing,String}=nothing,
+)::Result{Any,ParseError}
+  stream = JuliaSyntax.ParseStream(str)
+
+  JuliaSyntax.parse!(stream; rule)
+
+  if peek(stream, skip_newlines=false, skip_whitespace=false) != K"EndMarker"
+    JuliaSyntax.emit_diagnostic(stream, error="unexpected text after parsing $rule")
+  end
+
+  if JuliaSyntax.any_error(stream.diagnostics) || !JuliaSyntax.isempty(stream.diagnostics)
+    ex = JuliaSyntax.ParseError(stream; filename)
+    return ParseError(Any, str, "Failed to parse String into Julia expression", backtrace(), ex)
+  end
+  JuliaSyntax.build_tree(Expr, stream; filename)
+end
+
+parse_julia(
+  str::AbstractString,
+  rule::Symbol=:statement,
+  filename::Union{Nothing,String}=nothing,
+) = unwrap(safe_parse_julia(str, rule, filename))
+
+"""Alternative to `Meta.parseall` that does not return incomplete expressions and doesn't throw, but wraps returned value (or error) in a `Result`."""
+safe_parseall_julia(str::AbstractString, filename::Union{Nothing,String}=nothing) =
+  safe_parse_julia(str, :all, filename)
+
+"""Alternative to `Meta.parseall` that does not return incomplete expressions. Throws on error."""
+parseall_julia(str::AbstractString, filename::Union{Nothing,String}=nothing) = parse_julia(str, :all, filename)
+
+
+# ---------------------------
+# Evaluation
+
+
+struct EvalError <: Exception
+  mod::Module
+  expr::Union{Expr,Symbol}
+  exc::Exception
+  bt::Backtrace
+end
+
+format_code(ex::Symbol) = repr(ex)
+
+function Base.showerror(io::IO, err::EvalError)
+  expression_str = replace(repr(err.expr), r"^"m => "  ")
```

Comment:
Potentially misleading in output. Why do this?

Reply:
It's just for tabbing. `^` is start of line regex, i misread it as the char too. Why misleading? it just "tabs" the expression